core/src/scene/filters.cpp

```diff
@@ -1,11 +1,214 @@
 #include "filters.h"
 #include "scene/styleContext.h"
 #include "data/tileData.h"
+#include "platform.h"
 
 #include <cmath>
 
 namespace Tangram {
 
+void countTypes(const std::vector<Filter>& filters, int& global, int& function, int& property) {
+
+}
+
+void Filter::print(int _indent) const {
+
+    switch (data.get_type_index()) {
+
+    case Data::type<OperatorAny>::value: {
+        logMsg("%*s any\n", _indent, "");
+        for (const auto& filt : data.get<OperatorAny>().operands) {
+            filt.print(_indent + 2);
+        }
+        break;
+    }
+    case Data::type<OperatorAll>::value: {
+        logMsg("%*s all\n", _indent, "");
+        for (const auto& filt : data.get<OperatorAll>().operands) {
+            filt.print(_indent + 2);
+        }
+        break;
+    }
+    case Data::type<OperatorNone>::value: {
+        logMsg("%*s none\n", _indent, "");
+        for (const auto& filt : data.get<OperatorNone>().operands) {
+            filt.print(_indent + 2);
+        }
+        break;
+    }
+    case Data::type<Existence>::value: {
+        auto& f = data.get<Existence>();
+        logMsg("%*s existence - key:%s\n", _indent, "", f.key.c_str());
+        break;
+    }
+    case Data::type<Equality>::value: {
+        auto& f = data.get<Equality>();
+        if (f.values[0].is<std::string>()) {
+            logMsg("%*s equality - global:%d key:%s val:%s\n", _indent, "",
+                   f.global != FilterGlobal::undefined,
+                   f.key.c_str(),
+                   f.values[0].get<std::string>().c_str());
+        }
+        if (f.values[0].is<double>()) {
+            logMsg("%*s equality - global:%d key:%s val:%f\n", _indent, "",
+                   f.global != FilterGlobal::undefined,
+                   f.key.c_str(),
+                   f.values[0].get<double>());
+        }
+        break;
+    }
+    case Data::type<Range>::value: {
+        auto& f = data.get<Range>();
+        logMsg("%*s range - global:%d key:%s min:%f max:%f\n", _indent, "",
+               f.global != FilterGlobal::undefined,
+               f.key.c_str(), f.min, f.max);
+        return;
+    }
+    case Data::type<Function>::value: {
+        logMsg("%*s function\n", _indent, "");
+        break;
+    }
+    default:
+        break;
+    }
+
+}
+
+
+int Filter::matchCost() const {
+    // Add some extra penalty for set vs simple filters
+    int sum = -100;
```

Comment:
The negative costing is feeling very hard to reason about, and seems like it could have bad edge cases; e.g. "set" filters are intended to have negative costs, but couldn't a "set" filter whose operands are "set" filters return a positive cost? Maybe I'm mistaken, but that would evidence my first point :)